source/common/filesystem/filesystem_impl.cc

```diff
@@ -153,12 +92,14 @@ FileImpl::~FileImpl() {
   }
 
   // Flush any remaining data. If file was not opened for some reason, skip flushing part.
-  if (fd_ != -1) {
+  if (raw_file_->isOpen()) {
     if (flush_buffer_.length() > 0) {
       doWrite(flush_buffer_);
     }
 
-    os_sys_calls_.close(fd_);
+    const auto result = raw_file_->close();
+    ASSERT(result.rc_, fmt::format("unable to close file '{}': {}", raw_file_->path(),
+                                   strerror(result.errno_)));
```

Comment:
IMO we should be abstracting this need to do stderror() in the RawFile interface.